Add optional string support to native_functions schema

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44877 WIP: torch.fft helper functions
* #44876 Deprecate old fft functions
* #44550 torch.fft: Multi-dimensional transforms
* #43011 Add one dimensional FFTs to torch.fft namespace
* **#43010 Add optional string support to native_functions schema**
* #43009 Cleanup tracer handling of optional arguments

Differential Revision: [D23751851](https://our.internmc.facebook.com/intern/diff/D23751851)

peterbell10 committed:
Add optional string support to native_functions schema

[ghstack-poisoned]

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

bhosmer commented on aten/src/ATen/native_parse.py:
```diff
@@ -85,6 +85,8 @@ def type_argument_translations(arg):
     elif t == 'str':
         t = 'std::string'
+    elif t == 'str?':
+        t = 'std::string?'
```
I think you'll need to add the inverse conversion [here](https://github.com/pytorch/pytorch/blob/master/tools/jit/gen_unboxing_wrappers.py#L43-L77).

peterbell10 replied:
Thanks, added.

bhosmer reviewed (commented):
Looks good! Need to teach codegen in one more place (noted inline) and then I think you're good to go.

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

bhosmer reviewed (approved):
Looks good, thanks!

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

anjali411 commented on tools/jit/gen_unboxing_wrappers.py:
```diff
@@ -135,6 +136,7 @@ def jit_type_of(arg):
     'int64_t': '{}.toInt()',
     'int64_t?': '{}.toOptional<int64_t>()',
     'std::string': '{}.toStringRef()',
+    'std::string?': '{}.toOptional<std::string>()',
```
Following the mapping for `std::string`, I think it should be`'std::string?': '{}.toOptionalStringRef()'`

peterbell10 replied:
After some experimenting, I think that would require the C++ signature of native functions be changed from `c10::optional<std::string>` to `c10::optional<std::reference_wrapper<const std::string>>` which I think is really ugly.

A nicer way to avoid the copy would be to change from `std::string` to `c10::string_view` everywhere, in which case we don't need the reference_wrapper.

peterbell10 replied:
But I figure the copy is okay since `std::string` functions will take a copy currently anyway. So I think a `string_view` conversion could be left for a separate PR.

anjali411 commented on torch/csrc/jit/frontend/tracer.h:
```diff
@@ -262,6 +262,10 @@ TORCH_API void addInputs(
+TORCH_API void addInputs(
+    Node* n,
+    const char* name,
+    const c10::optional<std::string>& value);
```
`const c10::optional<std::reference_wrapper<const std::string>> value`

anjali411 commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -186,6 +186,7 @@ struct PythonArgs {
   inline c10::optional<at::MemoryFormat> memoryformatOptional(int i);
   inline at::QScheme toQScheme(int i);
   inline std::string string(int i);
+  inline c10::optional<std::string> stringOptional(int i);
```
`inline c10::optional<std::reference_wrapper<const std::string>> stringOptional(int i);`

anjali411 commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -520,6 +521,11 @@ inline std::string PythonArgs::string(int i) {
   return THPUtils_unpackString(args[i]);
 }
 
+inline c10::optional<std::string> PythonArgs::stringOptional(int i) {
```
`inline c10::optional<std::reference_wrapper<const std::string>> PythonArgs::stringOptional(int i) {`

peterbell10 replied:
That would cause a dangling reference.

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add optional string support to native_functions schema"




[ghstack-poisoned]

peterbell10 committed:
Update on "Add optional string support to native_functions schema"


Differential Revision: [D23751851](https://our.internmc.facebook.com/intern/diff/D23751851)

[ghstack-poisoned]
